### ISSUE TITLE
Segfault in tree iterator

### ISSUE THREAD

**Issue by Therzok** [libgit2/libgit2#3616]
Whenever trying to issue a blame operation or a log operation on a specific file, I get a crash.

I haven't tried a bisect yet, but https://github.com/libgit2/libgit2/commit/1947592461a7f6b97fa32f4517ba65f89ff428e0 is the last known good, and master is the known bad.

https://github.com/libgit2/libgit2/commit/3679ebaef5436a662ad74819f6cbd2c1e76e6766 doesn't seem to fix it either.

Repro steps - Windows only, I haven't tested other OS:
1. Checkout https://github.com/mono/monodevelop .
2. Use the blame example in libgit2 on this path `main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitRepository.cs`
3. Observe segfault with stacktrace below:
```
 	git2.dll!tree_iterator__pop_frame(tree_iterator * ti, unsigned char final) Line 561	C
 	git2.dll!tree_iterator__pop_all(tree_iterator * ti, unsigned char to_end, unsigned char final) Line 581	C
 	git2.dll!tree_iterator__free(git_iterator * self) Line 778	C
 	git2.dll!git_iterator_free(git_iterator * iter) Line 1870	C
 	git2.dll!git_iterator_for_tree(git_iterator * * iter, git_tree * tree, git_iterator_options * options) Line 836	C
 	git2.dll!git_diff_tree_to_tree(git_diff * * diff, git_repository * repo, git_tree * old_tree, git_tree * new_tree, const git_diff_options * opts) Line 1334	C
 	git2.dll!find_origin(git_blame * blame, git_commit * parent, git_blame__origin * origin) Line 434	C
 	git2.dll!pass_blame(git_blame * blame, git_blame__origin * origin, unsigned int opt) Line 529	C
 	git2.dll!git_blame__like_git(git_blame * blame, unsigned int opt) Line 624	C
 	git2.dll!blame_internal(git_blame * blame) Line 334	C
 	git2.dll!git_blame_file(git_blame * * out, git_repository * repo, const char * path, git_blame_options * options) Line 373	C
 	blame.exe!main(int argc, char * * argv) Line 80	C
 	blame.exe!invoke_main() Line 74	C++
 	blame.exe!__scrt_common_main_seh() Line 264	C++
 	blame.exe!__scrt_common_main() Line 309	C++
 	blame.exe!mainCRTStartup() Line 17	C++
 	kernel32.dll!@BaseThreadInitThunk@12()	Unknown
 	ntdll.dll!__RtlUserThreadStart()	Unknown
 	ntdll.dll!__RtlUserThreadStart@8()	Unknown
```

`tf` ends up `NULL` and my libgit2-fu is not enough to track it down by myself.

Sorry for the crude repro steps, but this is all I have to reproduce it. I know it's a consistent repro with that file, so it's all I got.

Other info:
* cmake args: `-DSTDCALL=ON -G "Visual Studio 14" -DBUILD_EXAMPLES=ON`
* I haven't seen this happen on small files.

**Comment by pks-t** [comment]
I wasn't able to reproduce this error on Linux. Following your hint that this only happens with big files I tried to blame the biggest file (StronglyTypedProperties.cs) in the MonoDevelop repository, but no crash here either. So I presume it might be Windows-specific.

**Comment by carlosmn** [comment]
The Windows code will often run as 32-bit, and we use fairly different codepaths to get the filesystem data loaded, so testing on any other OS likely won't yield many results.

This looks like it might be the same as #3585

**Comment by Therzok** [comment]
Confirmed, that patch fixes the issue.

**Comment by carlosmn** [comment]
Fixed via #3619
